crypto/hpke.c

```diff
@@ -0,0 +1,2730 @@
+ * Or if you'd like to re-caclulate the sha256 of nothing...
+ *  SHA256_CTX sha256;
+ *  SHA256_Init(&sha256);
+ *  char* buffer = NULL;
+ *  int bytesRead = 0;
+ *  SHA256_Update(&sha256, buffer, bytesRead);
+ *  SHA256_Final(zero_sha256, &sha256);
+ * ...but I've done it for you, so no need:-)
+ * static const unsigned char zero_sha256[SHA256_DIGEST_LENGTH] = {
+ *     0xe3, 0xb0, 0xc4, 0x42, 0x98, 0xfc, 0x1c, 0x14,
+ *     0x9a, 0xfb, 0xf4, 0xc8, 0x99, 0x6f, 0xb9, 0x24,
+ *     0x27, 0xae, 0x41, 0xe4, 0x64, 0x9b, 0x93, 0x4c,
+ *     0xa4, 0x95, 0x99, 0x1b, 0x78, 0x52, 0xb8, 0x55};
+ * </pre>
+ */
+
+/*!
+ * @brief decode ascii hex to a binary buffer
+ *
+ * @param ahlen is the ascii hex string length
+ * @param ah is the ascii hex string
+ * @param blen is a pointer to the returned binary length
+ * @param buf is a pointer to the internally allocated binary buffer
+ * @return 1 for good otherwise bad
+ */
+int hpke_ah_decode(
+        size_t ahlen, const char *ah,
+        size_t *blen, unsigned char **buf)
+{
+    size_t lblen=0;
+    size_t i=0;
+    unsigned char *lbuf=NULL;
+    if (ahlen <=0 || ah==NULL || blen==NULL || buf==NULL) {
+        return 0;
+    }
+    if (ahlen%1) {
+        return 0;
+    }
+    lblen=ahlen/2;
+    lbuf=OPENSSL_malloc(lblen);
+    if (lbuf==NULL) {
+        return 0;
+    }
+    for (i=0;i!=lblen;i++) {
+        lbuf[i]=HPKE_A2B(ah[2*i])*16+HPKE_A2B(ah[2*i+1]);
+    }
+    *blen=lblen;
+    *buf=lbuf;
+    return 1;
+}
+
+#if defined(SUPERVERBOSE) || defined(TESTVECTORS)
+/*!
+ * @brief for odd/occasional debugging
+ *
+ * @param fout is a FILE * to use
+ * @param msg is prepended to print
+ * @param buf is the buffer to print
+ * @param blen is the length of the buffer
+ * @return 1 for success
+ */
+static int hpke_pbuf(FILE *fout, char *msg,unsigned char *buf,size_t blen)
+{
+    if (!fout) {
+        return 0;
+    }
+    if (!msg) {
+        fprintf(fout,"NULL msg:");
+    } else {
+        fprintf(fout,"%s (%lu): ",msg,blen);
+    }
+    if (!buf) {
+        fprintf(fout,"buf is NULL, so probably something wrong\n");
+        return 1;
+    }
+    if (blen==HPKE_MAXSIZE) {
+        fprintf(fout,"length is HPKE_MAXSIZE, so probably unused\n");
+        return 1;
+    }
+    if (blen==0) {
+        fprintf(fout,"length is 0, so probably something wrong\n");
+        return 1;
+    }
+    size_t i=0;
+    for (i=0;i<blen;i++) {
+        fprintf(fout,"%02x",buf[i]);
+    }
+    fprintf(fout,"\n");
+    return 1;
+}
+#endif
+
+/*!
+ * @brief Check if kem_id is ok/known to us
+ * @param kem_id is the externally supplied kem_id
+ * @return 1 for good, not-1 for error
+ */
+static int hpke_kem_id_check(uint16_t kem_id)
+{
+    switch (kem_id) {
+        case HPKE_KEM_ID_P256:
+        case HPKE_KEM_ID_P384:
+        case HPKE_KEM_ID_P521:
+        case HPKE_KEM_ID_25519:
+        case HPKE_KEM_ID_448:
+            break;
+        default:
+            return(__LINE__);
```

Comment:
strange:  `return(__LINE__);` do not return `-1`

Reply:
Left that in for now.